### PR TITLE
Add thread safety annotations to xds_end2end_test

### PR DESCRIPTION
**Pull request by tamird**
Follow up to #26867.

@yashykt

### CODE REVIEW

**Review by yashykt — APPROVED** [review]

File: test/cpp/end2end/xds_end2end_test.cc
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:842]: This doesn't look right to me.  `ProcessRequest()` is called while already holding the lock, so I think this will deadlock.

I think you just need to tag this method with `ABSL_EXCLUSIVE_LOCKS_REQUIRED`.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:946]: Same problem here: `ProcessUpdate()` is called when already holding the lock, so this should deadlock.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:1004]: Same here.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: test/cpp/end2end/xds_end2end_test.cc
Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: You're absolutely right. Done.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: Done.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: Done.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc:647]: Looks like this one is also called only when holding the lock, so this is also a deadline.  This method also needs `ABSL_EXCLUSIVE_LOCKS_REQUIRED()`.

Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: Yup. Done.

**Comment by markdroth** [comment]
@veblush Any idea why `ABSL_EXCLUSIVE_LOCKS_REQUIRED()` doesn't seem to be working here?

**Comment by tamird** [comment]
I added some `ABSL_ASSERT_EXCLUSIVE_LOCK` and explanatory comments. This smells like a compiler bug to me - I would be happy to report it upstream but don't think it should hold up this PR.

I've also added a few more annotations that I missed the first time around.

**Comment by tamird** [comment]
Something is fishy about the fact that the compiler errors mention the template instantiation. There's an attempted reduction in https://gcc.godbolt.org/z/3P17KbvhP which seems to work correctly even when the annotation goes through a pointer, so something else is going on. Again, looks like a compiler bug.

File: test/cpp/end2end/xds_end2end_test.cc
Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc:842]: That seems strictly worse. Why throw out the baby with the bathwater? This comment is just a breadcrumb for future readers, and this lambda is a very narrow workaround, where `NO_THREAD_SAFETY_ANALYSIS` is throwing the guardrails out entirely.

 I am currently working on producing a small repro that can be sent upstream as a bug report.

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc:842]: I'd really like to hear back from upstream as to what's actually causing this problem before we merge this PR.  I want to make sure we're not somehow misunderstanding something.

Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc:842]: Well, this is looking quite likely to be a compiler bug at this point. https://bugs.llvm.org/show_bug.cgi?id=51368

Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: I've updated this workaround to be more precise. It will now cause compilation to *fail* if the compiler bug is fixed, which is good because it will prompt a human to come around and remove the workaround.

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc]: Actually, that may not be ideal, because when the new compiler is released, we'll need the code to work with both the old and new compilers.  Can we do this in a way that will bypass the bug but continue to work when the bug is fixed?

Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: We certainly can, but do we want to? My worry is that we will simply forget about this code. How about a compromise: I'll add a snippet in this comment that explains what the compiler appeasing alternative would be (probably just a call to AssertHeld) so that the compiler upgrading person can do so.

This won't affect downstream users because this is test code.

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc]: If the problem is that the compiler isn't able to translate the lock annotations through the template, is there a way that we can make it work via a proxy object *inside* the template?  For example, maybe we could create our a new mutex class inside the template that just delegates all methods to `parent_->ads_mu_`, and then we can have the lock annotations refer to that lock instead of referring directly to `parent_->ads_mu_`?

Comment (markdroth) [reply test/cpp/end2end/xds_end2end_test.cc]: To be clear, I'm really not happy with `AssertHeld()`, because that becomes a run-time error instead of a compile-time error.

Comment (tamird) [reply test/cpp/end2end/xds_end2end_test.cc]: This suggestion won't work because the compiler will then not be able to see that accessing guarded state through `parent_` is permitted.

**Comment by markdroth** [comment]
This looks good.  I've triggered the tests, so let's see if everything is happy.

**Review by markdroth — APPROVED** [review]
Just waiting for the last couple of tests to complete.

**Comment by markdroth** [comment]
Known issues: #26933 #26595

The "Basic Tests Ruby MacOS" and "Artifact Build MacOS" failures are infrastructure timeouts.

**Pull request merged**
